compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/nodes/gc/shenandoah/ShenandoahBarrierSet.java

```diff
@@ -0,0 +1,342 @@
+            }
+            case AbstractCompareAndSwapNode cmpSwap -> {
+                if (useCASBarrier) {
+                    addWriteBarriers(cmpSwap, cmpSwap.getNewValue(), cmpSwap.getExpectedValue(), false);
+                    if (cmpSwap instanceof ValueCompareAndSwapNode) {
+                        addReadNodeBarriers(cmpSwap);
+                    }
+                }
+            }
+            case ArrayRangeWrite ignored -> GraalError.unimplemented("ArrayRangeWrite is not used");
+            case null, default ->
+                GraalError.guarantee(n.getBarrierType() == BarrierType.NONE, "missed a node that requires a GC barrier: %s", n.getClass());
+        }
+    }
+
+    private void addWriteBarriers(FixedAccessNode node, ValueNode writtenValue, ValueNode expectedValue, boolean doLoad) {
+        BarrierType barrierType = node.getBarrierType();
+        switch (barrierType) {
+            case NONE:
+                // nothing to do
+                break;
+            case FIELD:
+            case ARRAY:
+            case UNKNOWN:
+            case POST_INIT_WRITE:
+            case AS_NO_KEEPALIVE_WRITE:
+                if (isObjectValue(writtenValue)) {
+                    StructuredGraph graph = node.graph();
+                    boolean init = node.getLocationIdentity().isInit();
+                    if (!init && barrierType != BarrierType.AS_NO_KEEPALIVE_WRITE && useSATBBarrier) {
+                        // The pre barrier does nothing if the value being read is null, so it can
+                        // be explicitly skipped when this is an initializing store.
+                        // No keep-alive means no need for the pre-barrier.
+                        addShenandoahPreWriteBarrier(node, node.getAddress(), expectedValue, doLoad, graph);
+                    }
+                    if (!init && useCardBarrier && !StampTool.isPointerAlwaysNull(writtenValue)) {
+                        graph.addAfterFixed(node, graph.add(new ShenandoahCardBarrierNode(node.getAddress())));
+                    }
+                }
+                break;
+            default:
+                throw new GraalError("unexpected barrier type: " + barrierType);
+        }
+    }
+
+    private void addLoadReferenceBarrier(FixedAccessNode node, AddressNode address, BarrierType barrierType) {
+        GraalError.guarantee(node != null, "input value must not be null");
+        StructuredGraph graph = node.graph();
+        boolean narrow = node.stamp(NodeView.DEFAULT) instanceof NarrowOopStamp;
+        ValueNode uncompressed = maybeUncompressReference(node, narrow);
+        ShenandoahLoadRefBarrierNode lrb = graph.add(new ShenandoahLoadRefBarrierNode(uncompressed, address, barrierType, narrow));
+        ValueNode compValue = maybeCompressReference(lrb, narrow);
+        ValueNode newUsage = uncompressed != node ? uncompressed : lrb;
+        node.replaceAtUsages(compValue, InputType.Value, usage -> usage != newUsage);
```

Comment:
I think this can just be `node.replaceAtUsages(compValue, usage -> usage != newUsage);`.  compValue is a regular Java value so it can only have `Value` usages.  So you don't need the changes in Node.java either, which I wasn't a fan of.

Reply:
I am not sure what lead me to think checking for the InputType was necessary, tbh. Maybe during development something else was wrong elsewhere and I would see non-Value input for some reason. Anyhow, I changed/reverted that part. Thanks for pointing out!